### PR TITLE
Teacher app: Rescore taught in past

### PR DESCRIPTION
**Pull request by clareconstantine**
Rescore the question about what CS courses you've taught previously to give bonus points if their answer doesn't include anything except:
- CS Fundamentals
- Robotics
- ScratchEd
- CS in Algebra
- CS in Science
- I don't have experience teaching any...

There is a script to rescore previous applications and produce a list of the changed applications sorted by regional partner for Anthonette to notify partners.

### CODE REVIEW

**Review by hacodeorg — APPROVED** [review]

File: dashboard/app/models/pd/application/teacher1920_application.rb [hunk 1]
```diff
@@ -505,7 +505,7 @@ def auto_score!
         meets_scholarship_criteria_scores[:previous_yearlong_cdo_pd] = responses[:previous_yearlong_cdo_pd].exclude?('CS Principles') ? YES : NO
       end
 
-      bonus_points_scores[:taught_in_past] = responses[:taught_in_past] == [options[:taught_in_past].last] ? 2 : 0
+      bonus_points_scores[:taught_in_past] = (responses[:taught_in_past] & options[:taught_in_past].values_at(3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 15)).any? ? 0 : 2
```
Comment (Hamms) [inline dashboard/app/models/pd/application/teacher1920_application.rb]: Relying on the indexes here seems unnecessarily fragile. Could this be pulled out into a helper method that's more explicit about the values we care about?

File: bin/oneoff/rescore_taught_in_past [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  if old_taught_in_past_score != new_taught_in_past_score
+    if application.update(response_scores: response_scores)
+      partner_name = application.regional_partner&.name
+      name = "#{application.first_name} #{application.last_name}"
+      update_string = "partner: #{partner_name}, id: #{application.id}, email: #{application.email}, name: #{name}"
+      rescored_applications << update_string
+
+      # append to the file as we go, so that in case we encounter an exception and exit early,
+      # we have a record of the applications we've updated.
+      File.open('rescored_applications', 'a') do |list|
```
Comment (Hamms) [inline bin/oneoff/rescore_taught_in_past]: nit; this should probably be in `/tmp/`

File: bin/oneoff/rescore_taught_in_past [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ActiveRecord::Base.transaction do
+  rescored_applications = []
+  puts "Rescoring taught_in_past for teacher 1920 applications..."
+  Pd::Application::Teacher1920Application.find_each do |application|
+    responses = application.sanitize_form_data_hash
+    options = application.class.options
+    response_scores = JSON.parse(application.response_scores)
+
+    old_taught_in_past_score = response_scores["bonus_points_scores"] ? response_scores["bonus_points_scores"]["taught_in_past"] : nil
+    new_taught_in_past_score = (responses[:taught_in_past] & options[:taught_in_past].values_at(3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 15)).any? ? 0 : 2
```
Comment (Hamms) [inline bin/oneoff/rescore_taught_in_past]: nit: this should probably also use `TAUGHT_IN_PAST_NO_BONUS_POINT_RESPONSES`, although fragility for a one-off script is much less frightening

File: bin/oneoff/rescore_taught_in_past [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        update_string = "partner: #{partner_name}, id: #{application.id}, old: #{old_taught_in_past_score}, new: #{new_taught_in_past_score}, response: #{responses[:taught_in_past]}, email: #{application.email}, name: #{name}"
+        rescored_applications << update_string
+      end
+    end
+  rescue => e
+    puts "Failure on application #{application.id}:\n#{e.message}\n#{application&.errors&.messages}"
+    raise
+  end
+
+  File.open('./tmp/rescored_applications', 'w') do |file|
```
Comment (Hamms) [inline bin/oneoff/rescore_taught_in_past:34]: why `./tmp` rather than `/tmp`?

**Review by Hamms — APPROVED** [review]

**Pull request merged**